pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotInstanceRestletResource.java

```diff
@@ -0,0 +1,204 @@
+
+  @GET
+  @Path("/instances")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "List all instances")
+  @ApiResponses(value = {@ApiResponse(code=200, message = "Success"),
+      @ApiResponse(code=500, message = "Error reading instances")})
+  public Instances getAllInstances(
+  ) {
+    return new Instances(pinotHelixResourceManager.getAllInstances());
+  }
+
+  @GET
+  @Path("/instances/{instanceName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Get instance information", produces = MediaType.APPLICATION_JSON)
+  @ApiResponses(value = {@ApiResponse(code=200, message = "Success"),
+      @ApiResponse(code = 404, message = "Instance not found"),
+      @ApiResponse(code=500, message = "Error reading instances")})
+  public String getInstance(
+      @ApiParam(value = "Instance name", required = true,
+          example = "Server_a.b.com_20000 | Broker_my.broker.com_30000")
+      @PathParam("instanceName") String instanceName)
+  {
+
+    if (!pinotHelixResourceManager.instanceExists(instanceName)) {
+      throw new WebApplicationException("Instance " + instanceName + " does not exist",
+          javax.ws.rs.core.Response.Status.NOT_FOUND);
+    }
+    InstanceConfig instanceConfig = pinotHelixResourceManager.getHelixInstanceConfig(instanceName);
+    JSONObject response = new JSONObject();
+    try {
+      response.put("instanceName", instanceConfig.getInstanceName());
+      response.put("hostName", instanceConfig.getHostName());
+      response.put("enabled", instanceConfig.getInstanceEnabled());
+      response.put("port", instanceConfig.getPort());
+      response.put("tags", new JSONArray(instanceConfig.getTags()));
+    } catch (JSONException e) {
+      throw new WebApplicationException(e, Response.Status.INTERNAL_SERVER_ERROR);
+    }
+    return response.toString();
+  }
+
+  @POST
+  @Path("/instances")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "Create a new instance", consumes = MediaType.APPLICATION_JSON, notes = "Creates a new instance with given instance config")
+  @ApiResponses(value = {@ApiResponse(code = 200, message="Instance successfully created"),
+      @ApiResponse(code=409, message="Instance exists already"),
+      @ApiResponse(code = 500, message = "Internal error")})
+  public SuccessResponse addInstance(
+      Instance instance
+  ) {
+    LOGGER.info("Instance creation request received for instance " + instance.toInstanceId());
+    final PinotResourceManagerResponse resp = pinotHelixResourceManager.addInstance(instance);
+    if (resp.status == PinotResourceManagerResponse.ResponseStatus.failure) {
+      throw new WebApplicationException("Instance exists already",
+          javax.ws.rs.core.Response.Status.CONFLICT);
+    }
+    return new SuccessResponse("Instance successfully created");
+  }
+
+  // TODO: @consumes text/plain but swagger doc says json. Does that work. It's better this way if it works
```

Comment:
@antumbde, can you please expand on this comment (since you added it)? IIRC if I changed it to json it did not work. we can leave the TODO and experiment later.